src/ucp/api/ucp.h

```diff
@@ -2177,8 +2177,112 @@ ucs_status_t ucp_worker_fence(ucp_worker_h worker);
+ *
+ * This enumeration defines the type of atomic memory operation should be 
+ * performed by the ucp_atomic_post family of fuctions. All of these are
+ * non-fetching atomics and will not result in a request handle.
+ */
+typedef enum {
+    UCP_ATOMIC_POST_OP_ADD, /**< Atomic add */
+    UCP_ATOMIC_POST_OP_LAST,
+} ucp_atomic_post_op_t;
 
 /**
+ * @ingroup UCP_COMM
+ * @brief Atomic operation requested for ucp_atomic_fetch
+ *
+ * This enumeration defines which atomic memory operation should be performed
+ * by the ucp_atomic_fetch family of functions. All of these functions
+ * will fetch data from the remote node
+ */
+typedef enum {
+    UCP_ATOMIC_FETCH_OP_FADD, /**< Atomic Fetch and add */
+    UCP_ATOMIC_FETCH_OP_SWAP, /**< Atomic swap */
+    UCP_ATOMIC_FETCH_OP_CSWAP, /**< Atomic conditional swap */
+    UCP_ATOMIC_FETCH_OP_LAST,
+} ucp_atomic_fetch_op_t;
+
+/**
+ * @ingroup UCP_COMM
+ * @brief Post an atomic memory operation.
+ *
+ * This routine posts an atomic memory operation to a remote value.
+ * The remote value is described by the combination of the remote
+ * memory address @a remote_addr and the @ref ucp_rkey_h "remote memory handle"
+ * @a rkey.
+ * Return from the function does not guarantee completion. A user must
+ * call @ref ucp_ep_flush or @ref ucp_worker_flush to guarentee that the remote
+ * value has been updated.
+ *
+ * @param [in] ep          UCP endpoint.
+ * @param [in] opcode      One of @ref ucp_atomic_post_op_t.
+ * @param [in] value       Source operand for the atomic operation.
+ * @param [in] op_size     Size of value in bytes
+ * @param [in] remote_addr Remote address to operate on.
+ * @param [in] rkey        Remote key handle for the remote address.
+ *
+ * @return Error code as defined by @ref ucs_status_t
+ */
+ucs_status_t ucp_atomic_post(ucp_ep_h ep, ucp_atomic_post_op_t opcode, uint64_t value,
+                             size_t op_size, uint64_t remote_addr, ucp_rkey_h rkey);
+
+/**
+ * @ingroup UCP_COMM
+ * @brief Post an atomic fetch operation.
+ *
+ * This routine will post an atomic fetch operation to remote memory.
+ * The remote value is described by the combination of the remote
+ * memory address @a remote_addr and the @ref ucp_rkey_h "remote memory handle"
+ * @a rkey.
+ * The routine is non-blocking and therefore returns immediately. However the 
+ * actual atomic operation may be delayed. The atomic operation is not considered complete
+ * until the values in remote and local memory are completed. If the atomic operation 
+ * operation is completed immediately the routine returns UCS_OK and the 
+ * call-back routine @a cb is @b not invoked. If the operation is @b not 
+ * completed immediately and no error reported then the UCP library will 
+ * schedule to invoke the call-back routine @a cb whenever the atomic operation will be
+ * completed. In other words, the completion of an atomic operation can be signaled by 
+ * the return code or the call-back.
```

Comment:
I suggest the following re-wording:
If the atomic operation completes immediately, the routine returns UCS_OK and the
call-back routine @a cb is @b not invoked.  If the operation is @b not completed
immediately and no error is reported, then the UCP library will schedule invocation of
the call-back routine @a cb upon completion of the atomic operation. In other words,
the completion of an atomic operation can be signaled by the return code or execution
of the call-back.